File: 1462_course_schedule_iv.py
class Solution(object):
    def checkIfPrerequisite(self, numCourses, prerequisites, queries):
        reach = [[False] * numCourses for _ in range(numCourses)]
        for u, v in prerequisites:
            reach[u][v] = True
        for k in range(numCourses):
            for i in range(numCourses):
                for j in range(numCourses):
                    reach[i][j] = reach[i][j] or (reach[i][k] and reach[k][j])
        return [reach[u][v] for u, v in queries]

File: 1462_course_schedule_iv.java
class Solution {
    public List<Boolean> checkIfPrerequisite(int c, int[][] p, int[][] q) {
        boolean[][] d = new boolean[c][c];
        for (int[] r : p) d[r[0]][r[1]] = true;
        for (int a = 0; a < c; a++)
            for (int b = 0; b < c; b++)
                for (int e = 0; e < c; e++)
                    d[b][e] |= d[b][a] && d[a][e];
        List<Boolean> ans = new ArrayList<>();
        for (int[] f : q) ans.add(d[f[0]][f[1]]);
        return ans;
    }
}
